### PR TITLE
Add vendor configuration support for Chartbeat analytics

### PR DESCRIPTION
**Pull request by britice**
Addresses issue https://github.com/ampproject/amphtml/issues/1436.

This implementation is dependent upon PR https://github.com/ampproject/amphtml/pull/1702 for interval requests.

An outstanding issue is the engaged time metrics (as described in issue https://github.com/ampproject/amphtml/issues/1296) which are needed to have a more complete implementation.

### CODE REVIEW

File: extensions/amp-analytics/0.1/vendors.js [hunk 1]
```diff
@@ -107,6 +107,44 @@ export const ANALYTICS_CONFIG = {
...
+      'trackInterval': {
+        'on': 'timer',
+        'timer-spec': {
+          'interval': 15
+        },
+        'request': 'interval'
+      }
+    },
+    'transport': {
+      'beacon': false,
```
Comment (britice) [inline extensions/amp-analytics/0.1/vendors.js]: Currently, Chartbeat only supports image requests but we are looking into adding support for `sendBeacon` requests.

Comment (cramforce) [inline extensions/amp-analytics/0.1/vendors.js]: How do you tell the difference? It should look the same on your side.

Comment (britice) [inline extensions/amp-analytics/0.1/vendors.js]: `sendBeacon` is a `POST` request. Our setup only accept `GET` requests.

File: extensions/amp-analytics/amp-analytics.md [hunk 2]
```diff
@@ -62,6 +62,7 @@ when the document is first loaded, and each time an `<a>` tag is clicked:
   - `type` This optional attribute can be specified to use one of the built-in analytics providers. Currently supported values for type are:
     - `googleanalytics`: Adds support for Google Analytics. More details for adding Google Analytics support can be found at [developers.google.com](https://developers.google.com/analytics/devguides/collection/amp-analytics/).
     - `comscore`: Supports comScore Unified Digital Measurement™ pageview analytics. Requires defining *var* `c2` with comScore-provided *c2 id*.
+    - `chartbeat`: Adds support for Chartbeat. More details for adding Chartbeat support can be found at [support.chartbeat.com](http://support.chartbeat.com/docs/).
```
Comment (avimehta) [inline extensions/amp-analytics/amp-analytics.md]: Please move this above googleanalytics. If you don't mind, please also move comscore above google analytics. t would be nice to have alphabetical ordering here.

File: extensions/amp-analytics/0.1/vendors.js [hunk 3]
```diff
@@ -107,6 +107,44 @@ export const ANALYTICS_CONFIG = {
       'xhrpost': false,
       'image': true
     }
+  },
+
+  'chartbeat': {
```
Comment (avimehta) [inline extensions/amp-analytics/0.1/vendors.js]: Please move this above google analytics. @rudygalfi mentioned that having alphabetical ordering would be good. I forgot to mention this to comscore guys so I'll fix their config but at least you can have it in the right place :)

Thanks!

File: extensions/amp-analytics/0.1/vendors.js [hunk 4]
```diff
@@ -107,6 +107,44 @@ export const ANALYTICS_CONFIG = {
...
+        'w=${screenHeight}&' +
+        'j=${decayTime}&' +
+        't=${pageViewId}&' +
+        'i=${title}&_',
+    },
+    'triggers': {
+      'trackInterval': {
+        'on': 'timer',
+        'timer-spec': {
+          'interval': 15
```
Comment (avimehta) [inline extensions/amp-analytics/0.1/vendors.js]: did you want to add a max-timer-interval value here?

File: extensions/amp-analytics/amp-analytics.md (hunk 2 above)
Comment (avimehta) [inline extensions/amp-analytics/amp-analytics.md]: having a link to amp-analytics specific documentation would help pubs a lot. If not, at least some page mentions AMP? If you are planning to update the link with AMP related info then this is fine.

File: extensions/amp-analytics/0.1/vendors.js [hunk 5]
```diff
@@ -107,6 +107,44 @@ export const ANALYTICS_CONFIG = {
...
+
+  'chartbeat': {
+    'vars': {
+      'decayTime': 15
+    },
+    'requests': {
+      'host': 'https://ping.chartbeat.net',
+      'interval': '${host}/ping?h=${domain}&' +
+        'p=${canonicalPath}&' +
+        'u=${clientId}&' +
```
Comment (avimehta) [inline extensions/amp-analytics/0.1/vendors.js]: clientId var expects a scope parameter. If you guys have a cookie that you use on publisher domains, having that cookie name would make AMP read that cookie if available. If you don't want to share cookies and want to just use some other namespace, please specify that namespace.

Example: `${clientId(chartbeat-cookie-name)}`

File: extensions/amp-analytics/0.1/vendors.js [hunk 6]
```diff
@@ -107,6 +107,44 @@ export const ANALYTICS_CONFIG = {
...
+        'd=${canonicalHost}&' +
+        'g=${uid}&' +
+        'g0=${sections}&' +
+        'g1=${authors}&' +
+        'g2=${zone}&' +
+        'g3=${sponsorName}&' +
+        'g4=${contentType}&' +
+        'x=${scrollTop}&' +
+        'w=${screenHeight}&' +
+        'j=${decayTime}&' +
```
Comment (avimehta) [inline extensions/amp-analytics/0.1/vendors.js]: Not sure what decay time is. Is this related to the timer trigger? Do you want to define the var there?

File: extensions/amp-analytics/0.1/vendors.js [hunk 7]
```diff
@@ -107,6 +107,44 @@ export const ANALYTICS_CONFIG = {
...
+      'decayTime': 15
+    },
+    'requests': {
+      'host': 'https://ping.chartbeat.net',
+      'interval': '${host}/ping?h=${domain}&' +
+        'p=${canonicalPath}&' +
+        'u=${clientId}&' +
+        'd=${canonicalHost}&' +
+        'g=${uid}&' +
+        'g0=${sections}&' +
```
Comment (avimehta) [inline extensions/amp-analytics/0.1/vendors.js]: g0-4 use vars that are not currently defined. Can you explain what those are? If you would like AMP to provide those, can you open an issue and cc @rudygalfi ? If you would like to send a PR for that, that would be even better!

**Comment by avimehta** [comment]
Mostly lgtm. Some minor comments. Thanks for sending out the PR!

File: extensions/amp-analytics/amp-analytics.md (hunk 2 above)
Comment (britice) [inline extensions/amp-analytics/amp-analytics.md]: > If you are planning to update the link with AMP related info then this is fine.

Yeah, we're working on internal documentation for this and will update the link when we have something up.

**Comment by cramforce** [comment]
I'm not 100% sure if all comments are addressed (or postponed). Happy to merge!

**Comment by britice** [comment]
@cramforce No, it's not quite ready. I'll likely have one more commit to address comments, another internal review, and then squash things back down to one commit.

File: extensions/amp-analytics/amp-analytics.md (hunk 2 above)
Comment (rudygalfi) [inline extensions/amp-analytics/amp-analytics.md]: Could you alphabetize this section as well?

**Comment by britice** [comment]
As a note, I will be pushing to get everything done on our end by sometime this afternoon (Eastern Time).

File: extensions/amp-analytics/0.1/vendors.js (hunk 7 above)
Comment (britice) [inline extensions/amp-analytics/0.1/vendors.js]: > If this information is available in JSON-LD, we could read it from there. Is that something you guys would be interested in?

Ultimately, yes that would be something we're interested in. However, at this point we are trying to add as little complexity for the publisher as possible. In our existing data collection functionality, these are values set specifically for the publisher's Chartbeat implementation. If we move the configuration away from the Chartbeat amp-analytics tag, there will likely be confusion with implementation for publishers with limited dev resources.

File: extensions/amp-analytics/0.1/vendors.js (hunk 6 above)
Comment (britice) [inline extensions/amp-analytics/0.1/vendors.js]: > Not sure what decay time is.

Decay time is the amount of time in seconds till our system should should assume that the page session has ended. This allows us to release data through the system more reliably. In our data collection JS, there is a concept of "back-off" which causes data to be sent less frequently if the user is not engaged with the content. That causes the decay time to increase which causes our system to wait longer for the next request to be sent.

> Is this related to the timer trigger? Do you want to define the var there?

Yeah, that makes sense.

**Comment by britice** [comment]
@avimehta, I pushed two more commits which should address all comments.

Is it preferred to have PRs consist of only one commit? If so, I can squash it down to a single commit as long as everything else looks good.

File: extensions/amp-analytics/0.1/vendors.js [hunk 8]
```diff
@@ -57,6 +57,80 @@ export const ANALYTICS_CONFIG = {
...
+          'interval': 15,
+          'max-timer-length': 7200
+        },
+        'request': 'interval',
+        'vars': {
+          'decayTime': 30
+        }
+      },
+      'trackAnchorClick': {
+        'on': 'click',
```
Comment (avimehta) [inline extensions/amp-analytics/0.1/vendors.js]: I am a little worried about this trigger because this will mean that anyone using charbeat will also start sending hits on `a` clicks. It's not a lot of overhead but wanted to bring @cramforce 's attention to this.

**Comment by avimehta** [comment]
The PR looks good to me. Please squash into a single commit. 

@cramforce, @rudygalfi  do you want to go over the PR once? Also, the travis failure seems to be because validation does not recognize chartbeat yet. Can we make validation read the recognized types from vendors.js? if not, how do you want to handle this?

File: extensions/amp-analytics/0.1/vendors.js (hunk 8 above)
Comment (britice) [inline extensions/amp-analytics/0.1/vendors.js]: Good point. This is meant to be a stop gap until there is a way to send data at the end of a page session. This will help to capture at least a portion of users that navigate away from the page.

**Comment by cramforce** [comment]
This likely will now fail with the error (for comscore, because it now validates unexpectedly). I already fixed that at head and will merge this PR if that is the only failure.

**Comment by cramforce** [comment]
Travis is down https://www.traviscistatus.com/

We will update later.

**Comment by cramforce** [comment]
And merging.
